### PR TITLE
[TACHYON-975] Enable standby masters to reuse state before becoming leader

### PR DESCRIPTION
**Pull request by gpang**
https://tachyon.atlassian.net/browse/TACHYON-975

### CODE REVIEW

File: servers/src/main/java/tachyon/master/MasterBase.java [hunk 1]
```diff
@@ -100,13 +100,24 @@ public void start(boolean isLeader) throws IOException {
...
         catchupTailer.processNextJournalLogFiles();
+      } else {
+        // This master has not successfully processed any of the journal, so create a fresh tailer
+        // to process the entire journal.
+        LOG.info(getServiceName() + ": process entire journal before becoming leader master.");
+        catchupTailer = new JournalTailer(this, mJournal);
+        if (catchupTailer.checkpointExists()) {
+          catchupTailer.processJournalCheckpoint(true);
+          catchupTailer.processNextJournalLogFiles();
+        }
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/MasterBase.java]: Shall we log the info that the checkpoint does not exist?

File: servers/src/main/java/tachyon/master/MasterBase.java [hunk 2]
```diff
@@ -125,7 +136,7 @@ public void start(boolean isLeader) throws IOException {
 
   @Override
   public void stop() throws IOException {
-    LOG.info(getServiceName() + ":Stopping master. isLeader: " + mIsLeader);
+    LOG.info(getServiceName() + ": Stopping master. isLeader: " + mIsLeader);
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/MasterBase.java]: What about constructing a more readable log:
mIsLeader?"leading":"shadow"

File: servers/src/main/java/tachyon/master/Master.java [hunk 3]
```diff
@@ -74,4 +75,13 @@
...
   void stop() throws IOException;
+
+  /**
+   * Provides the master with a {@link Journal} capable of writing, in preparation to starting as
+   * the leader. This enables transitioning from standby master to leader master without having to
+   * construct a new master object.
+   *
+   * @param journal the {@link Journal} capable of writing
+   */
+  void upgradeToWriteJournal(Journal journal);
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/Master.java]: This reads a little weird, given the hierarchy of journal. Right now the readonly journal extends read-write journal, we may consider the opposite: let read-write journal extend the readonly version.

**Comment by yupeng9** [comment]
Thanks for the improvement. I left some comments.

**Comment by uronce-cc** [comment]
LGTM

File: servers/src/main/java/tachyon/master/MasterBase.java (hunk 1 above)
Comment (gpang) [inline servers/src/main/java/tachyon/master/MasterBase.java]: done

File: servers/src/main/java/tachyon/master/MasterBase.java (hunk 2 above)
Comment (gpang) [inline servers/src/main/java/tachyon/master/MasterBase.java]: done

File: servers/src/main/java/tachyon/master/Master.java (hunk 3 above)
Comment (gpang) [inline servers/src/main/java/tachyon/master/Master.java]: That is a good point. How should that work? If the less functional read-only version is the base class, how do users of the base class know to get the write functionality? Does it have to do instance checks and casting?

Another option might be to have a journal interface, and both readonly and readwrite implement it.

**Comment by gpang** [comment]
Thanks for the reviews. I had a response/question about your review comment, @yupeng9 .

**Comment by gpang** [comment]
I restructured the journal class hierarchy.

**Comment by yupeng9** [comment]
LGTM

**Comment by yupeng9** [comment]
Thanks for the journal hierarchy change!

**Pull request merged**